distribution/lib/Standard/Database/0.0.0-dev/src/Extensions/Upload_Database_Table.enso

```diff
@@ -45,16 +43,29 @@ from project.Internal.Upload_Table import all
 
      If an error has been raised, the table is not created (that may not always
      apply to `SQL_Error`).
+
+   ? Dry Run
+
+     If the Output context is disabled, only a dry run is performed. This means
+     that the operation checks for errors like missing columns.
+
+     More expensive checks, like clashing keys are checked only on a sample of
+     1000 rows, so it is possible for the operation to encounter one of these
```

Comment:
I wonder if this optimization should be a separate flag; 'dry run' means 'no real side effects' rather than 'speed up expensive checks'. The two flags might often be used together, but perhaps they should be different.

Reply:
The dry run is not really a flag, but specifies the behaviour when the Output context is disabled.

The only scenario we could do is dry run with checking _all_ values or checking _none_ at all (then the actual check will happen on the proper execution).

I don't think we should be adding additional flags for this - as this parameter would not at all apply when the context is enabled.

We can reconsider if we want to run the check at all. I think checking a subset is a compromise that will allow us to catch some errors while retaining the dry run behaviour and not sabotaging the performance. Do you think we should instead not check at all in dry run mode? Or check all entries?

-----

I like checking all entries because it actually does a 'proper' dry run - it verifies the behaviour that will happen on actual execution. The only worry was that it may be too expensive.

In the in memory backend, this operation will be as expensive as any operations preparing the data before hand (roughly O(N) cost).

Problem is a bit bigger in the DB backend, as all operations are done 'lazily' by default - they just construct more and more complex SQL queries, but do not run them. Now this check would require to run the query. Still, the cost is comparable to the cost of attaching a Table visualization to any of the queries.

And then, there is the issue that while `update_database_table` does the check, but does not actually retain the data; `select_into_database_table` is meant to create a temporary dry run table. If we create it with all the data, it's not much different from the 'proper' run, only difference is table name and that it is a temporary table. Still processing all the data gives us the closest experience to the actual run, apart from side effects.

-----

I like processing all data in the dry runs. I would consider trying it and going back to smaller samples if the performance is really unsatisfactory. @jdunkerley what do you think?